app/uk/gov/hmrc/tai/model/hip/reads/RateBandHipReads.scala

```diff
@@ -14,22 +14,27 @@
  * limitations under the License.
  */
 
-package uk.gov.hmrc.tai.model.domain
+package uk.gov.hmrc.tai.model.hip.reads
 
-import play.api.libs.json.{JsArray, JsValue}
+import play.api.libs.json.*
+import uk.gov.hmrc.tai.model.domain.RateBand
 
 object RateBandHipReads {
-  def incomeAndRateBands(json: JsValue): Seq[RateBand] = {
 
-    val bands = (json \ "totalLiabilityDetails" \ "nonSavings" \ "taxBandDetails").asOpt[JsArray]
-    val details = bands.map(_.value.collect {
-      case js if (js \ "income").asOpt[BigDecimal].isDefined =>
-        RateBand((js \ "income").as[BigDecimal], (js \ "rate").as[BigDecimal])
-    })
+  def incomeAndRateBands(json: JsValue): Seq[RateBand] = {
+    val taxBandPath = __ \ "totalLiabilityDetails" \ "nonSavings" \ "taxBandDetails"
 
-    details match {
-      case Some(rateBands) => rateBands.toSeq.sortBy(-_.rate)
-      case None            => Seq.empty[RateBand]
-    }
+    taxBandPath
+      .readNullable[Seq[JsValue]]
+      .reads(json)
+      .getOrElse(None)
+      .getOrElse(Seq.empty)
+      .flatMap { js =>
+        for {
+          income <- (js \ "income").asOpt[BigDecimal]
+          rate   <- (js \ "rate").asOpt[BigDecimal]
+        } yield RateBand(income, rate)
+      }
+      .sortBy(-_.rate)
```

Comment:
way too convoluted.

json is: 
```"taxBandDetails": [
        {
          "bandType": "B",
          "taxCode": "BR",
          "income": 6807,
          "tax": 1361.4,
          "lowerBand": 0,
          "upperBand": 31865,
          "rate": 20
        },
        {
          "bandType": "D0",
          "taxCode": "D0",
          "lowerBand": 31865,
          "upperBand": 150000,
          "rate": 40
        },
        {
          "bandType": "D1",
          "taxCode": "D1",
          "lowerBand": 150000,
          "upperBand": 0,
          "rate": 45
        }
      ]
      ```
      
      Should something like readNullable[List[RateBand]]
      In RateBand some implicit reads for rate and income

Reply:
Much clearer now